core/sdk-core/src/main/java/software/amazon/awssdk/core/retry/conditions/RetryCondition.java

```diff
@@ -32,14 +30,31 @@
      */
     boolean shouldRetry(RetryPolicyContext context);
 
+    /**
+     * Called by the SDK to notify this condition that the provided request will not be retried, because some retry condition
+     * determined that it shouldn't be retried.
+     */
+    default void requestWillNotBeRetried(RetryPolicyContext context) {
```

Comment:
Do we really need additional hooks? I feel there are some overlaps between them and interceptors. I'd be fine if this is internal, but this class is public api.

My concern is that having two places to inject `afterAttempt` hooks can be confusing.

Reply:
Why wouldn't this hook be needed? Implementing a stateful retry condition like a token bucket needs to know when the capacity it acquired can be released.

The only other alternatives, I don't like:
1. Don't make retry throttling directly configurable.
2. Remove `requestWillNotBeRetried` (the other one is still required), but we'd have to add validation that: (1) there's only one stateful retry condition, and (2) that stateful retry condition is always considered after every other retry condition. That feels more confusing, and we'd still need something like `isStatefulRetryCondition` (or more likely, a `StatefulRetryCondition` interface) to be able to identify which ones are stateful so that we could implement those preconditions.
3. Do an `instanceof ExecutionInterceptor` on every registered retry condition that's added so that we can also register them as an execution interceptor (somehow, not sure how that would work). We'd then implement `ExecutionInterceptor` on the retry conditions that need the information.